curriculum/challenges/english/20-upcoming-python/learn-string-manipulation-by-building-a-cipher/6555e5991af57d1ae0e35f0a.md

```diff
@@ -0,0 +1,114 @@
+---
+id: 6555e5991af57d1ae0e35f0a
+title: Step 91
+challengeType: 20
+dashedName: step-91
+---
+
+# --description--
+
+Wait a minute! You cannot decrypt anything with the wrong key. Try with `happycoding`.
+
+With that, your cipher project is complete.
+
+# --hints--
+
+You should still have a `custom_key` variable.
+
+```js
+({ test: () => assert(__userGlobals.has("custom_key")) })
+```
+
+You should modify the `custom_key` value into `happycoding`.
+
+```js
+({ test: () => assert.equal(__userGlobals.get("custom_key"), "happycoding") })
+```
+
+# --seed--
+
+## --seed-contents--
+
+```py
+--fcc-editable-region--
+text = 'mrttaqrhknsw ih puggrur'
+custom_key = 'python'
+
+def vigenere(message, key, direction=1):
+    key_index = 0
+    alphabet = 'abcdefghijklmnopqrstuvwxyz'
+    final_message = ''
+
+    for char in message.lower():
+
+        # Append any non-letter character to the message
+        if not char.isalpha():
+            final_message += char
+
+        else:        
+            # Find the right key character to encode/decode
+            key_char = key[key_index % len(key)]
+            key_index += 1
+
+            # Define the offset and the encrypted/decrypted letter
+            offset = alphabet.index(key_char)
+            index = alphabet.find(char)    
+            new_index = (index + offset*direction) % len(alphabet)
+            final_message += alphabet[new_index]
+    
+    return final_message
+
+def encrypt(message, key):
+    return vigenere(message, key)
+    
+def decrypt(message, key):
+    return vigenere(message, key, -1)
+
+print(f'\nEncrypted text: {text}')
+print(f'Key: {custom_key}')
+decryption = decrypt(text, custom_key)
+print(f'\nDecrypted text: {decryption}\n')
+--fcc-editable-region--
```

Comment:
```suggestion
```

Reply:
Here, I intentionally left the editable region wide since this is the last step.